packages/html/src/eval.rs

```diff
@@ -0,0 +1,93 @@
+use async_channel::Receiver;
+use async_trait::async_trait;
+use dioxus_core::ScopeState;
+use std::rc::Rc;
+
+/// A struct that implements EvalProvider is sent through [`ScopeState`]'s provide_context function
+/// so that [`use_eval`] can provide a platform agnostic interface for evaluating JavaScript code.
+pub trait EvalProvider {
+    fn new_evaluator(&self, cx: &ScopeState, js: String) -> Box<dyn Evaluator>;
+}
+
+/// The platform's evaluator.
+#[async_trait(?Send)]
+pub trait Evaluator {
+    /// Runs the evaluated JavaScript.
+    fn run(&mut self) -> Result<(), EvalError>;
+    /// Sends a message to the evaluated JavaScript.
+    fn send(&self, data: serde_json::Value) -> Result<(), EvalError>;
+    /// Gets an UnboundedReceiver to receive messages from the evaluated JavaScript.
+    fn receiver(&mut self) -> Receiver<serde_json::Value>;
+    /// Cleans up any evaluation artifacts.
+    fn done(&mut self);
+}
+
+/// Get a struct that can execute any JavaScript.
+///
+/// # Safety
+///
+/// Please be very careful with this function. A script with too many dynamic
+/// parts is practically asking for a hacker to find an XSS vulnerability in
+/// it. **This applies especially to web targets, where the JavaScript context
+/// has access to most, if not all of your application data.**
+pub fn use_eval<S: ToString>(cx: &ScopeState, js: S) -> &mut UseEval {
```

Comment:
Returning &mut UseEval here where eval is not clone makes it [difficult to move around UseEval](https://dioxuslabs.com/docs/nightly/guide/en/interactivity/custom_hooks.html#hook-anti-patterns). I think you should still be able to start an eval within a future with this API, but you would need to make the javascript await for a start message after you run it which could be unintuitive

I also wonder if we could make the API a bit more similar to the previous API to make upgrading less breaking. What are your thoughts on this API?
- use_eval returns a Rc<dyn EvalProvider>
- You can call `.start` on the eval provider to start running the script
- The Eval returned from `.start` can either be turned into a future with [IntoFuture](https://doc.rust-lang.org/nightly/core/future/trait.IntoFuture.html) (which consumes it) like the old API or you can use the send/receive function like you currently implemented

Reply:
I agree with the Rc but instead of a start function, how about implementing ``IntoFuture`` for the  ``UseEval`` struct? It would automatically run and await the first data returned through the channel. 

Additionally, any ``return data;`` JS could be replaced with ``dioxus.send(data);``? - it would add some string processing overhead during runtime though. If not that then we could add it as a step in the migration process - tell people to replace ``return`` with ``dioxus.send``?

```rs
// I could see this being useful outside of legacy code
let eval = use_eval(cx, "dioxus.send('hi from js');");
if let Ok(msg) = eval.await {
    // act on message
}
```
compared to the old:
```rs
let eval = use_eval(cx);
if let Ok(msg) = eval("return 'hi from js';").await {
    // act on message
}
```

Reply:
> I agree with the Rc but instead of a start function, how about implementing `IntoFuture` for the `UseEval` struct? It would automatically run and await the first data returned through the channel.

This sounds good!

> Additionally, any `return data;` JS could be replaced with `dioxus.send(data);`? - it would add some string processing overhead during runtime though. (**On second thoughts, replacing the return statement could have false positives**)

Nested function make this difficult. I think it would be better to try to not modify the javascript script submitted (except the wrapping we already do).

> If not that then we could add it as a step in the migration process - tell people to replace `return` with `dioxus.send`?
> 
> ```rust
> // I could see this being useful outside of legacy code
> let eval = use_eval(cx, "dioxus.send('hi from js');");
> if let Ok(msg) = eval.await {
>     // act on message
> }
> ```
> 
> compared to the old:
> 
> ```rust
> let eval = use_eval(cx);
> if let Ok(msg) = eval("return 'hi from js';").await {
>     // act on message
> }
> ```

I could see the old system being useful when you have a chain of several different scripts like:

```rust
let eval = use_eval(cx);
use_future!(cx, |()| {
    to_owned![eval];
    async move {
        eval("return 'script 1'").await;
        // some rust processing code
        eval("return 'script 2'").await;
    }
});
```
Where script 1 and script 2 are from different libraries.

In the new system, that would need to look like this:

```rust
let eval1 = use_eval(cx, "script 1");
let eval2 = use_eval(cx, "script 2");
use_future!(cx, |()| {
    to_owned![eval1, eval2];
    async move {
        eval1.await;
        // some rust processing code
        eval2.await;
    }
});
```

If we force the code to be defined with the hook it could also hurt colocation:

```rust
let script = use_eval(cx, "script 1");

// 100 more lines of code setting up data

render!{
    button {
        onclick: move |_| async move {
            script.await
        }
    }
}
```

Looking at the onclick handler, it could be less clear what code script.await is running compared to:

```rust
let eval = use_eval(cx);

// 100 more lines of code setting up data

render!{
    button {
        onclick: move |_| async move {
            eval("script 1").await
        }
    }
}
```

Reply:
> Looking at the onclick handler, it could be less clear what code script.await is running compared to: 
> ```rs
> let eval = use_eval(cx);
> 
> // 100 more lines of code setting up data
> 
> render!{
>     button {
>         onclick: move |_| async move {
>             eval("script 1").await
>         }
>     }
> }
> ```

I believe this shouldn't be as big of an issue with good variable naming. Developers could also move the use_eval call below the 100 lines.

> In the new system, that would need to look like this:
> ```rs
> let eval1 = use_eval(cx, "script 1");
> let eval2 = use_eval(cx, "script 2");
> use_future!(cx, |()| {
>     to_owned![eval1, eval2];
>     async move {
>         eval1.await;
>         // some rust processing code
>         eval2.await;
>     }
> });
> ```
I agree, chaining eval is useful. However, what if something like this was implemented which could effectively do the same:
```rs
let eval = use_eval(cx, 
    r#"
    script 1
    // wait for signal to continue
    await dioxus.signal(); // reads as "await dioxus signal"
    script 2
    "#);

use_future!(cx, |()| {
    to_owned![eval];
    async move {
        eval.run();
        // some rust processing code
        eval.signal(); // Signal the script to continue
    }
});
```
The dev could still pass data to the script when needed using the regular ``send()`` and ``recv()``.

I generally want to avoid using any closures or anonymous functions. In my opinion, they add complexity and make control flow a little confusing (dioxus-std was a pain).

Reply:
I did the stuff @Demonthos
This works
```rs
let mut eval = use_eval(cx);

// 100 more lines of code setting up data

render!{
    button {
        onclick: move |_| async move {
            eval("script 1").await
        }
    }
}
```